### PR TITLE
EventsExtension: fix compatibility with Nette 2.4

### PR DESCRIPTION
**Pull request by mrtnzlml**
Closes: https://github.com/Kdyby/Events/issues/99

### CODE REVIEW

**Comment by enumag** [comment]
@mrtnzlml It's not so easy, that method does not exist in Nette 2.3.

**Comment by mrtnzlml** [comment]
Oh, I see. I did not realize that.

**Comment by mrtnzlml** [comment]
Fixed.

What is the right way to run all tests using tester? I tried this, but without success:

screenshot from 2016-06-07 16 52 27

File: src/Kdyby/Events/DI/EventsExtension.php [hunk 1]
```diff
@@ -313,8 +313,14 @@ private function autowireEvents(Nette\DI\ContainerBuilder $builder)
 					continue;
 				}
 			}
-			if ($def->getImplementType() === 'get') {
-				continue;
+			if (method_exists($def, 'getImplementMode')) { //BC (Nette 2.4)
```
Comment (enumag) [inline src/Kdyby/Events/DI/EventsExtension.php]: Nette 2.4 is new so "BC" comment is wrong here imho. Also I'd rather put the detection outside of the foreach. It's not really important in compile time but still.

**Comment by enumag** [comment]
@mrtnzlml I'm not sure I understand - if you want to run all tests, just don't specify any (remove the last argument of your command).

**Comment by enumag** [comment]
What do you think about this instead?

```php
$implementGetter = method_exists('Nette\DI\ServiceDefinition', 'getImplementMode') ? 'getImplementMode' : 'getImplementType';
...
if ($def->{$implementGetter}() === 'get') {
```

**Comment by mrtnzlml** [comment]
@enumag Irrelevant. Same behavior...

About your solution - I don't like it, but you are the boss here, so should I change it?

**Comment by Majkl578** [comment]
Dynamic getter is by magnitude worse than method_exists in foreach imho. You can always declare boolean variable outside foreach and do a simple if check inside foreach.

**Comment by enumag** [comment]
@Majkl578 I know. It just seemed more readable.

@mrtnzlml Actually I'm not the boss here, that's @fprochazka of course. I don't have push rights to this repo.

The fastest solution is probably this:

```php
$newApi = method_exists('Nette\DI\ServiceDefinition', 'getImplementMode');
...
if (($newApi ? $def->getImplementMode() : $def->getImplementType()) === 'get') {
```

**Comment by mrtnzlml** [comment]
I know @enumag. And tu be honest, I don't really care. But now I am quite confused. Should I use the last proposed solution? If so, I'll do rebase...

**Comment by mrtnzlml** [comment]
@enumag Rebased. What about now?

**Comment by Majkl578** [comment]
Btw there's also new constant for this, `ServiceDefinition::IMPLEMENT_MODE_GET`, so how about this?
```php
$newApi = defined('Nette\DI\ServiceDefinition::IMPLEMENT_MODE_GET');
if ($newApi ? $def->getImplementMode() === $def::IMPLEMENT_MODE_GET : $def->getImplementType() === 'get')

**Comment by enumag** [comment]
@Majkl578 Nice I didn't know about the constant.

**Comment by mrtnzlml** [comment]
Me neither. I suppose this is faster (maybe)? I am getting tired of rebases. Every solution except the first one worked just fine. Please chose the Chosen One and I'll do rebase (again)... :)

**Comment by enumag** [comment]
You don't need rebase, just amend last commit.

I don't think it's my call but Majkl's last suggestion is good imo so personally I'd go with that.

**Comment by mrtnzlml** [comment]
@enumag I am using rebase so often, that I completely forgot about `--amend`. Thanks. But this was not the point... :)

Anyway, amended and green. Still cannot use `vender/bin/tester` with tests, because it doesn't work for some reason, but whatever...

**Comment by enumag** [comment]
@mrtnzlml Thanks!

@fprochazka Ready to merge in my opinion.

**Comment by Majkl578** [comment]
Hmm, tests are indeed a bit strange...
```
$ vendor/bin/tester -p php7.0-local tests/ | tail -n +6 | grep -vE '^\s*$'
PHP 7.0.7 | php7.0-local | 8 threads
...........................................................
OK (59 tests, 0.8 seconds)

$ vendor/bin/tester -p php7.0 tests/ | tail -n +6 | grep -vE '^\s*$'
PHP 7.0.7-4 | php7.0 | 8 threads
FFFFFFF
-- FAILED: Kdyby\Events\Event. | KdybyTests/Events/Event.phpt
-- FAILED: Kdyby\Events\EventArgs. | KdybyTests/Events/EventArgs.phpt
-- FAILED: Kdyby\Events\EventManager. | KdybyTests/Events/EventManager.phpt
-- FAILED: Kdyby\Events\Extension. | KdybyTests/Events/Extension.phpt
-- FAILED: Kdyby\Events\IExceptionHandler. | KdybyTests/Events/IExceptionHandler.phpt
-- FAILED: Kdyby\Events\LazyEventManager. | KdybyTests/Events/LazyEventManager.phpt
-- FAILED: Kdyby\Events\NamespacedEventManager. | KdybyTests/Events/NamespacedEventManager.phpt
FAILURES! (7 tests, 7 failures, 0.1 seconds)

$ vendor/bin/tester -p php5.6 tests/ | tail -n +6 | grep -vE '^\s*$'
PHP 5.6.22-2+0~20160601121249.1+jessie~1.gbp701a45 | php5.6 | 8 threads
FFFFFFF
-- FAILED: Kdyby\Events\Event. | KdybyTests/Events/Event.phpt
-- FAILED: Kdyby\Events\EventArgs. | KdybyTests/Events/EventArgs.phpt
-- FAILED: Kdyby\Events\EventManager. | KdybyTests/Events/EventManager.phpt
-- FAILED: Kdyby\Events\Extension. | KdybyTests/Events/Extension.phpt
-- FAILED: Kdyby\Events\IExceptionHandler. | KdybyTests/Events/IExceptionHandler.phpt
-- FAILED: Kdyby\Events\LazyEventManager. | KdybyTests/Events/LazyEventManager.phpt
-- FAILED: Kdyby\Events\NamespacedEventManager. | KdybyTests/Events/NamespacedEventManager.phpt
FAILURES! (7 tests, 7 failures, 0.1 seconds)

$ vendor/bin/tester -p php5.5 tests/ | tail -n +6 | grep -vE '^\s*$'
PHP 5.5.36-2+0~20160601121612.1+jessie~1.gbp08f27b | php5.5 | 8 threads
FFFFFFF
-- FAILED: Kdyby\Events\Event. | KdybyTests/Events/Event.phpt
-- FAILED: Kdyby\Events\EventArgs. | KdybyTests/Events/EventArgs.phpt
-- FAILED: Kdyby\Events\EventManager. | KdybyTests/Events/EventManager.phpt
-- FAILED: Kdyby\Events\Extension. | KdybyTests/Events/Extension.phpt
-- FAILED: Kdyby\Events\IExceptionHandler. | KdybyTests/Events/IExceptionHandler.phpt
-- FAILED: Kdyby\Events\LazyEventManager. | KdybyTests/Events/LazyEventManager.phpt
-- FAILED: Kdyby\Events\NamespacedEventManager. | KdybyTests/Events/NamespacedEventManager.phpt
FAILURES! (7 tests, 7 failures, 0.1 seconds)
```

**Comment by mrtnzlml** [comment]
> Hmm, tests are indeed a bit strange... But it might be a missing extension in my case (no idea which one though?).

@Majkl578 Good point! In my case it's always JSON, so `vendor/bin/tester tests/ -d extension=json.so` works fine... :) But the error is little bit strange.

**Comment by Majkl578** [comment]
Yep, I needed json.so and tokenizer.so.

**Comment by foxycode** [comment]
@fprochazka Will you please merge it and make new release?

**Comment by fprochazka** [comment]
Thank you guys

**Pull request merged**